### DISCUSSION TITLE
'SPI running at half speed' - other causes than SD card and possible fixes?

### DISCUSSION THREAD

**Discussion by fdanapfel** [erichelgeson/BlueSCSI-v1#175] [Q&A]
Hello,

I have two BlueSCSI devices (both were bought preassembled from the same vendor (OneGeekArmy)), one is the 1.1-a DB25 "PowerBook" version with the DB-25 connector, which works flawlessly.

The other is a 1.1-a "Desktop" version with the internal 50-pin connector, which always shows the warning about 'SPI running at half' speed, no matter which SD card I put in it:
```
BlueSCSI <-> SD - https://github.com/erichelgeson/BlueSCSI
VER: 1.1-20220917-USB
DEBUG:0 SDFAT_FILE_TYPE:3
SdFat version: 2.2.0
Sd Format: exFAT
SPI speed: 25Mhz
SPI running at half speed - read https://github.com/erichelgeson/BlueSCSI/wiki/Slow-SPI
SdFat Max FileName Length: 64
Initialized SD Card - let's go!
Sd MID:3 OID:SD
Sd Name:SD64G
Sd Date:6/2021
Sd Serial:421969346
 - HD00_512.hda HDD / 1003484160bytes / 979965KiB / 956MiB
 - HD10_512.hda HDD / 524288000bytes / 512000KiB / 500MiB
ID:LUN0:
 0: 512:
 1: 512:
 2:----:
 3:----:
 4:----:
 5:----:
 6:----:
Finished initialization of SCSI Devices - Entering main loop.

```

When the same SD card is used in the external BlueSSCSI device it works without showing this message in the log. 

I've followed all the tips in the mentioned link, and as you can see also tried updating the firmware to the latest version (the 'BlueSCSI-Updater' also works without issues on macOS 10.15.7 (Catalina) without issues btw.). 

But no matter what I try the message about SPI running at half speed doesn't go away.

Could this be an issue with the hardware itself? Is this actually something to worry about?

**Comment by erichelgeson** [comment]
The message is normally due to a poor connection on the SD card or poor quality SD cards (hence the guidance in the wiki article) - It really only slightly affects performance, but it is annoying to have. Just a thought, how are you powering the 50 pin? Have you measured the 5v? Please try powering it via USB (even just outside the computer) then look at the log and see if you get the same message. Also put some IPA on the card and run it in our out a few times - I've had some flux get down in there before and cause an issue like this. Hope that helps, let us know!

**Comment by fdanapfel** [comment]
Thanks for the response. Unfortunately none of the suggestion helped.

I'm using the "Berg" connector to get power from one of the standard hard disk power connectors on the various systems I've tried using the BlueSCSI so far, and even tried putting it in an external SCSI enclosure with its own power supply which didn't make a difference. But even when I power the BlueSCSI via USB I still get the message in the log.txt.

Where would I have to measure the 5V?

Btw. I've also tried cleaning the contacts with IPA multiple times without success.

Edit: I've now measured the Voltage between the Pins marked "G" and "5V" on the BluePill and the multimeter shows a value of 5.08V between those pins. I've also checked the Voltage of the SD Card slot between pins 3 and 5 and there I get a value of 3.29V, which I guess should be close enough to the required 3.3V for the SD card slot.

**Reply by erichelgeson** [reply]
Voltage seems fine. I should note that even at half speed the SPI bus is still faster than we push the SCSI bus - so while you may see a few kb/sec slower transfer - it really shouldn't affect anything. Maybe I should change the warning message to be a little less worrisome. 

In reality I've only ever seen this happen a handful of times so that's why our troubleshooting doesn't have more in it besides the few things we tried that did work for those few cases.

**Reply by erichelgeson** [reply]
[firmware.zip](https://github.com/erichelgeson/BlueSCSI/files/9720326/firmware.zip)
Can you give this firmware a try and see what the SPI is successful in the log? It's USB flashable - Use QMK toolbox https://github.com/qmk/qmk_toolbox as an easy way to flash.

**Reply by fdanapfel** [reply]
Can this firmware also be used on a BlueSCSI that has the 96MHz version of the BluePill installed (OneGeekArmy kindly swapped out the original 50pin BlueSCSI against one with the faster BluePill to see if it makes a difference with regards to the SPi issue)?

And could I also use the BlueSCSI-Updater to install this firmware version (since I've used it so far without issues on macOS 10.15.7 (Catalina))?

**Reply by erichelgeson** [reply]
[firmware96mhz.zip](https://github.com/erichelgeson/BlueSCSI/files/9725480/firmware96mhz.zip)

Here's a 96mhz build - while you can use the updater you'll have to edit the scripts to point to this file. I plan to switch to qmk going forward as it's much more user friendly.

**Reply by fdanapfel** [reply]
Thanks for the 96mhz build. I was able to flash the BlueSCSI using QMK Toolbox, which apparently works exactly the same as with the BlueSCSI-updater :)

Here is the log.txt after putting on the new firmware:
```
BlueSCSI <-> SD - https://github.com/erichelgeson/BlueSCSI
VER: 1.1-20220918-SNAPSHOT-USB-96MHz
DEBUG:0 SDFAT_FILE_TYPE:3
SdFat version: 2.2.0
Sd Format: exFAT
SPI speed: 35Mhz
SPI under 40Mhz - read https://github.com/erichelgeson/BlueSCSI/wiki/Slow-SPI
SdFat Max FileName Length: 64
Initialized SD Card - let's go!
Sd MID:13 OID:KG
Sd Name:SD08G
Sd Date:3/2013
Sd Serial:1627402910
 - HD20_512-1GB-MacPlus-6+7.hda HDD / 1073741824bytes / 1048576KiB / 1024MiB
ID:LUN0:
 0:----:
 1:----:
 2: 512:
 3:----:
 4:----:
 5:----:
 6:----:
Finished initialization of SCSI Devices - Entering main loop.
```
As you can see I used a different SD Card as before, but this SD Card works without issues in a ver1.1 DB25 BlueSCSI on the same computer.

**Comment by boodledog** [comment]
I have this issue too, slow SPI (25MHz) on a DB25 BlueSCSI from Bruce at RecapAMac. Tried 3x SD cards and no luck. 
I'll be asking Bruce tomorrow if he had the same issue, to see if it's on my end.

**Reply by erichelgeson** [reply]
Continued at https://tinkerdifferent.com/threads/slow-spi-half-speed-bluescsi-db25.1966/

**Reply by erichelgeson** [reply]
Seems like we've fixed this now with a firmware update - the fix will be in the next version. I've also updated the docs.
